crates/ruma-client-api/src/r0/keys/upload_signatures.rs

```diff
@@ -19,25 +24,76 @@ ruma_api! {
     request: {
         /// Signed keys.
         #[ruma_api(body)]
-        pub signed_keys: BTreeMap<Box<UserId>, BTreeMap<String, JsonValue>>,
+        pub signed_keys: BTreeMap<Box<UserId>, SignedKeys>,
     }
 
     #[derive(Default)]
-    response: {}
+    response: {
+        /// Signature processing failures.
+        pub failures: BTreeMap<Box<UserId>, BTreeMap<String, Failure>>,
+    }
 
     error: crate::Error
 }
 
 impl Request {
     /// Creates a new `Request` with the given signed keys.
-    pub fn new(signed_keys: BTreeMap<Box<UserId>, BTreeMap<String, JsonValue>>) -> Self {
+    pub fn new(signed_keys: BTreeMap<Box<UserId>, SignedKeys>) -> Self {
         Self { signed_keys }
     }
 }
 
 impl Response {
     /// Creates an empty `Response`.
     pub fn new() -> Self {
-        Self {}
+        Self::default()
+    }
+}
+
+/// A map of key IDs to signed key objects.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct SignedKeys(BTreeMap<Box<str>, Box<RawJsonValue>>);
+
+impl SignedKeys {
+    /// Creates an empty `SignedKeys` map.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add the given device keys.
+    pub fn add_device_keys(&mut self, device_id: Box<DeviceId>, device_keys: Raw<DeviceKeys>) {
+        self.0.insert(device_id.into(), device_keys.into_json());
     }
+
+    /// Add the given cross signing keys.
+    pub fn add_cross_signing_keys(
+        &mut self,
+        what_the_hell_is_this: Box<str>,
+        cross_signing_keys: Raw<CrossSigningKey>,
```

Comment:
Also why did we name this type `CrossSigningKey` instead of `CrossSigningKeys` (extra 's' at the end)? It has a `keys` field just like `DeviceKeys`.

Reply:
I don't think I named these so can't tell for sure, but probably because `DeviceKeys` will have two identity keys attached to the struct, a ed25519 key and a curve25519 key.

While there are multiple cross signing keys (master, user-signing, self-signing), each of those structs will contain only one of them.

The cross signing triplet represents a single cryptographic user identity, not sure why it isn't structured in such a way in Matrix land, perhaps to allow different use-cases in the future. Then we would have a `DeviceKeys` struct and a `UserKeys` struct and both of them would have multiple keys inside of them.

Reply:
Okay so the current naming makes sense for how things are specced? I'll leave it as-is then.

Reply:
Probably, there might be multiple master keys in the future since all the keys are inside a map, but the spec doesn't allow this currently.